### PR TITLE
Revert: Revert: Clean up CRD conversion code structure

### PR DESCRIPTION
**Pull request by sttts**
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

Refactor the CRD conversion code:

* Rename crConverter to delegatingCRConverter
* Make creating a delegatingCRConverter a package function NewDelegatingConverter
* Move list/single item handling from the webhook converter into delegatingCRConverter
* Rename crConverterInterface to CRConverter so it is exported
* Change the CRConverter interface's input/output to be *unstructured.UnstructuredList
* Create and export NewNOPConverter

This is another attempt on @ncdc's https://github.com/kubernetes/kubernetes/pull/113151 that got reverted in https://github.com/kubernetes/kubernetes/pull/117301 because it ~~caused~~ triggered https://github.com/kubernetes/kubernetes/issues/117260. Although the root cause (objects to be converted without metadata) seem to fixed in SSA (at least the integration test with the reproducer is green), the conversion code is made resilient to that situation too. Compare https://github.com/kubernetes/kubernetes/pull/120920#issuecomment-1739367993.

```release-note
NONE
```

### CODE REVIEW

**Comment by sttts** [comment]
Infra flake 

/retest

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/converter.go [hunk 1]
```diff
@@ -245,17 +251,26 @@ func (c *delegatingCRConverter) ConvertToVersion(in runtime.Object, target runti
...
+	// Deepcopy because the converter might mutate the objects.
+	for i := range objectsToConvert {
+		objectsToConvert[i] = *objectsToConvert[i].DeepCopy()
+	}
+
+	// Do the (potentially mutating) conversion.
+	convertedObjects, err := c.converter.Convert(&unstructured.UnstructuredList{
+		Object: list.Object,
+		Items:  objectsToConvert,
+	}, toGVK.GroupVersion())
```
Comment (sttts) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/converter.go:263]: @liggitt with lots of assumptions around the CRConverter interface we could allow mutation and make use of it by not deepcopying furtherdown. Not sure it is worth it.

**Comment by ncdc** [comment]
Thanks for resurrecting this @sttts! I'll try to review later today. Worst case, my day is wide open tomorrow.

**Comment by sttts** [comment]
Copied from Slack:

Here are my findings:
- the added integration tests fails at b98659b before the revert
- it does not fail before this PR! and not after this PR.
- the object in question was a pruned object from SSA with apiVersion, kind and status only, no metadata!
I don't see the same kind of object in master sent to conversion. There is always metadata. My suspicion: SSA got fixed.
- With the integration test added here @apelisse  and colleagues might be able to bisect and understand what's going on. Maybe they know by heart what has changed.
- @ncdc only moved validation after webhook invocation into the generic delegatingCRConverter, also used for NopConversion after this PR. Hence, it triggered previously for NopConversion. Changes are high that 
@stefanprodan’s finding with flux would have shown too if they had webhook conversion in use.

**Comment by liggitt** [comment]
there are definitely paths that invoke conversion without metadata (see https://github.com/kubernetes/kubernetes/issues/117356#issuecomment-1516736527)

**Comment by sttts** [comment]
@liggitt have added logic to handle missing metadata. The other topic of partial objects and conversion is bigger cc @jpbetz @apelisse

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/converter.go [hunk 2]
```diff
@@ -396,47 +396,52 @@ func validateConvertedObject(in, out *unstructured.Unstructured) error {
...
+	originalMeta, ok := om.(map[string]interface{})
+	if !ok {
+		return fmt.Errorf("invalid metadata of type %T in input object", om)
 	}
 
-	obj = converted.Object["metadata"]
-	convertedMetaData, ok := obj.(map[string]interface{})
-	if !ok {
-		return fmt.Errorf("invalid metadata of type %T in input object", obj)
+	result := converted
```
Comment (ncdc) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/converter.go:419]: Is it worth having a shallow copy here? Might make things more confusing below?

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/converter.go [hunk 3]
```diff
@@ -396,47 +396,52 @@ func validateConvertedObject(in, out *unstructured.Unstructured) error {
...
+	if previouslyFound {
+		result.Object["metadata"] = originalMeta
+	} else {
+		result.Object["metadata"] = map[string]interface{}{}
 	}
```
[the remark this answers is not in the document]
Comment (sttts) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/converter.go:424]: good idea. Done.

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/converter.go (hunk 2 above)
Comment (sttts) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/converter.go:419]: done

**Comment by ncdc** [comment]
LGTM. @liggitt ?

**Comment by liggitt** [comment]
do you have a pointer to the `mutation and converter input bug` referenced in the commit? is that the same as the missing metadata bug? trying to understand what each commit is addressing

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/converter.go [hunk 4]
```diff
@@ -234,9 +233,8 @@ func (c *delegatingCRConverter) ConvertToVersion(in runtime.Object, target runti
 			return in, nil
 		}
 		// for a list, set the version of the top-level list object (all individual objects are already in the correct version)
-		out := list.DeepCopy()
```
Comment (liggitt) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/converter.go]: I'm trying to understand why the deep copy was there and if it's safe to remove... looks like it was original to the converter in https://github.com/kubernetes/kubernetes/pull/67006

I can't tell if this is introducing modification of the input (or if we were already doing that), and if so, if that's ok. I see the existing function docs around converting `in` in-place, but I'm not sure callers actually had that expectation if it wasn't happening previously.

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/converter.go [hunk 5]
```diff
@@ -252,17 +250,34 @@ func (c *delegatingCRConverter) ConvertToVersion(in runtime.Object, target runti
 		return converted, nil
 	}
 
-	convertedObjects, err := c.converter.Convert(list, toGVK.GroupVersion())
+	// Deepcopy ObjectMeta because the converter might mutate the objects, and we
+	// need the original ObjectMeta furthermore for validation and restoration.
+	for i := range objectsToConvert {
+		original := objectsToConvert[i].Object
+		objectsToConvert[i].Object = make(map[string]interface{}, len(original))
```
Comment (liggitt) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/converter.go:257]: similar questions about whether modifying the list input in-place is ok

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/converter.go [hunk 6]
```diff
@@ -396,47 +396,52 @@ func validateConvertedObject(in, out *unstructured.Unstructured) error {
 
 // restoreObjectMeta copies metadata from original into converted, while preserving labels and annotations from converted.
 func restoreObjectMeta(original, converted *unstructured.Unstructured) error {
```
Comment (liggitt) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/converter.go:398]: did the test added exercise the path where SSA invoked this on an object without metadata, and the conversion returned annotations?

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/converter.go (hunk 4 above)
Comment (sttts) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/converter.go]: the interface does not require the implementation to be safe, see above "Note that the function may mutate in object and return it." We wrap it with another deepcopy lateron.

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/converter.go (hunk 5 above)
Comment (sttts) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/converter.go:257]: same answer: "Note that the function may mutate in object and return it."

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/converter.go (hunk 6 above)
Comment (sttts) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/converter.go:398]: no, it didn't. Added.

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/webhook_converter_test.go [hunk 7]
```diff
@@ -61,7 +61,7 @@ func TestRestoreObjectMeta(t *testing.T) {
 		{"invalid original metadata",
 			map[string]interface{}{"metadata": []interface{}{"foo"}},
 			map[string]interface{}{"metadata": map[string]interface{}{}, "spec": map[string]interface{}{}},
-			map[string]interface{}{"metadata": []interface{}{"foo"}, "spec": map[string]interface{}{}},
+			map[string]interface{}{"metadata": map[string]interface{}{}, "spec": map[string]interface{}{}},
```
[the remark this answers is not in the document]
Comment (sttts) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/webhook_converter_test.go:64]: actually we don't do either. We return an error and converted is just unchanged.

**Comment by sttts** [comment]
As a follow-up, I wouldn't mind to switch the whole converter interface to "always safe". It reduces complexity and we are not making (much) use (if at all) of not having to deepcopy.

**Comment by sttts** [comment]
> do you have a pointer to the `mutation and converter input bug` referenced in the commit? is that the same as the missing metadata bug? trying to understand what each commit is addressing

@liggitt this was about bugs in `delegatingCRConverter.ConvertToVersion` of https://github.com/kubernetes/kubernetes/pull/113151/files#diff-dcc30cc8add91b73cae8d32919e61cfa3bdb965205d898656fd7f29652061dc0R217, e.g. that 

```golang
objectsToConvert, err := getObjectsToConvert(list, desiredAPIVersion, c.validVersions)
```

passed `list`, i.e. the full list of objects, not those to be converted. And with that `list` is potentially mutated, unexpectedly for the code further down.

**Comment by sttts** [comment]
@liggitt please take another look

**Comment by liggitt** [comment]
Queued up, sorry for the delay, got preempted by a lot of http/2 mitigation work this week.

I like the goal here of making the no-op conversion and webhook conversion consistent, but I'm wary of existing bugs the previous refactor exposed, which made bugs that only applied to the webhook conversion path *also* apply to the no-op conversion path.

The three areas I was trying to trace / verify didn't regress were:
* Performance
  * Does this change deep copy _more_ for either no-op conversion or webhook conversion?
* Contract about modification of inputs
  * Does this change newly modify input objects for either no-op conversion or webhook conversion?
  * Even if the documentation said that was a possibility, I'd want to be sure extensions-apiserver callers all worked properly when that happens, which I hadn't traced yet.
* Handling of conversion requests missing object metadata
  * The worst case scenario I can think of is 1) an object which has labels/annotations, 2) a code path (e.g. https://github.com/kubernetes/kubernetes/issues/117356) tries to convert with metadata stripped, and 3) the convertor also makes use of labels/annotations ... I don't really know what should happen in that case. I can't tell what this change does in that case.
  * I *think* the test case added covers issues reported like https://github.com/fluxcd/flux2/issues/3179 and https://github.com/Azure/azure-service-operator/issues/3343
  * I'm not sure if it covers cases of modifying labels/annotations (e.g. https://github.com/Azure/azure-service-operator/issues/3148) or the webhook convertor making use of labels/annotations
  * I actually think the root cause bug is SSA stripping metadata in conversion requests – if I could order changes, I would prefer to see that SSA bug fixed *before* trying to compensate for it here

**Comment by sttts** [comment]
@liggitt looking into deepcopies, and safe+unsafe converters. I don't think what we are doing makes a lot of sense. The nop converter is trivial to not need deepcopy. Top-level map (which includes apiVersion) is enough. And the webhook converter does marshalling/unmarshalling anyway. So there there deepcopy is pretty much automatic.

**Comment by erikgb** [comment]
Are there any new initiatives for fixing this? I just got hit by https://github.com/fluxcd/flux2/issues/3179. In my case the issue popped up after upgrading an operator where the `stored` version of an API (CRD) was bumped. The Flux server-side dry-run where all failing for resources using the previous `apiVersion`, probably because of conversion taking place as part of the dry-run. While resources using the new (and stored) `apiVersion` were not affected.

**Pull request closed without merging**